File: sdk/js/src/changes.mjs
/**
 * Changes detection: scan the upper directory to extract file modifications
 * relative to the base layer.
 *
 * Detects:
 *   - Added   files (in upper, not in base)
 *   - Modified files (in both layers)
 *   - Deleted  files (whiteout markers in upper)
 *
 * Supports both host-side whiteout format (.wh.<name> files) and
 * kernel overlay whiteout format (char device 0/0), including the reserved
 * `.wh..wh.*` metadata (opaque-dir markers, hardlink index) fuse-overlayfs
 * writes — it is the COW engine used on host mounts that cannot store kernel
 * overlay metadata (macOS/Docker virtiofs).
 */

import fs from 'node:fs';
import path from 'node:path';

const WH_PREFIX = '.wh.';
// Reserved OCI/fuse-overlayfs metadata namespace: `.wh..wh..opq` (opaque
// directory marker) and `.wh..wh.plnk` (hardlink index).  These are never
// user-visible entries — treating them as whiteouts for a file named `.wh.opq`
// / `.wh.plnk` produced phantom deletions on the fuse-overlayfs path (the COW
// engine boxsh falls back to when the kernel overlay cannot store overlay
// metadata, e.g. macOS/Docker virtiofs host mounts).
const WH_META_PREFIX = '.wh..wh.';
const WH_OPAQUE = '.wh..wh..opq';
const MANIFEST_DIR = '.boxsh';
const MANIFEST_SUFFIX = '.manifest';

/**
 * Scan the upper directory and return a list of all changes relative to base.
 *
 * @param {{ upper: string, base: string }} options
 * @returns {Array<{ path: string, type: 'added'|'modified'|'deleted' }>}
 */
export function getChanges(options) {
    const manifest = _readCloneManifest(options.upper);
    if (manifest) {
        return _getCloneSnapshotChanges(options.upper, options.base, manifest);
    }

    const changes = [];
    _scanChanges(options.upper, options.base, '.', changes);
    return changes.sort((a, b) => a.path.localeCompare(b.path));
}

function _getCloneSnapshotChanges(upperRoot, baseRoot, manifestPaths) {
    const changes = [];
    const deletedAncestors = new Set();

    for (const relPath of manifestPaths) {
        if (_hasDeletedAncestor(relPath, deletedAncestors)) continue;
        if (fs.existsSync(path.join(upperRoot, relPath))) continue;
        deletedAncestors.add(relPath);
        changes.push({ path: relPath, type: 'deleted' });
    }

    _scanCloneSnapshotChanges(upperRoot, baseRoot, '.', changes);
    return changes.sort((a, b) => a.path.localeCompare(b.path));
}

/**
 * @param {string} upperRoot
 * @param {string} baseRoot
 * @param {string} rel
 * @param {Array<object>} changes
 */
function _scanChanges(upperRoot, baseRoot, rel, changes) {
    const upperDir = path.join(upperRoot, rel);
    if (!fs.existsSync(upperDir) || !fs.statSync(upperDir).isDirectory()) return;

    for (const name of fs.readdirSync(upperDir)) {
        const childRel  = rel === '.' ? name : path.join(rel, name);
        const upperPath = path.join(upperRoot, childRel);

        // fuse-overlayfs metadata (opaque marker, hardlink index): never a
        // change entry of its own.  An opaque marker hides the lower content
        // of this directory, so the lower entries missing from the upper are
        // genuine deletions.
        if (name.startsWith(WH_META_PREFIX)) {
            if (name === WH_OPAQUE) {
                _collectOpaqueDeletions(upperRoot, baseRoot, rel, changes);
            }
            continue;
        }

        // Host-side whiteout (.wh.<name>)
        if (name.startsWith(WH_PREFIX)) {
            const targetName = name.slice(WH_PREFIX.length);
            const targetRel  = rel === '.' ? targetName : path.join(rel, targetName);
            changes.push({ path: targetRel, type: 'deleted' });
            continue;
        }

        // Kernel overlay whiteout (char device 0/0)
        if (_isKernelWhiteout(upperPath)) {
            changes.push({ path: childRel, type: 'deleted' });
            continue;
        }

        const st = fs.statSync(upperPath);

        if (st.isDirectory()) {
            const basePath = path.join(baseRoot, childRel);
            if (!fs.existsSync(basePath)) {
                changes.push({ path: childRel, type: 'added' });
            }
            _scanChanges(upperRoot, baseRoot, childRel, changes);
        } else if (st.isFile()) {
            const basePath = path.join(baseRoot, childRel);
            changes.push({ path: childRel, type: fs.existsSync(basePath) ? 'modified' : 'added' });
        }
    }
}

function _scanCloneSnapshotChanges(upperRoot, baseRoot, rel, changes) {
    const upperDir = path.join(upperRoot, rel);
    if (!fs.existsSync(upperDir) || !fs.statSync(upperDir).isDirectory()) return;

    for (const name of fs.readdirSync(upperDir)) {
        const childRel = rel === '.' ? name : path.join(rel, name);
        const upperPath = path.join(upperRoot, childRel);

        if (name.startsWith(WH_META_PREFIX)) {
            if (name === WH_OPAQUE) {
                _collectOpaqueDeletions(upperRoot, baseRoot, rel, changes);
            }
            continue;
        }

        if (name.startsWith(WH_PREFIX)) {
            const targetName = name.slice(WH_PREFIX.length);
            const targetRel = rel === '.' ? targetName : path.join(rel, targetName);
            changes.push({ path: targetRel, type: 'deleted' });
            continue;
        }

        if (_isKernelWhiteout(upperPath)) {
            changes.push({ path: childRel, type: 'deleted' });
            continue;
        }

        const upperStat = fs.lstatSync(upperPath);
        const basePath = path.join(baseRoot, childRel);

        if (upperStat.isDirectory()) {
            if (!fs.existsSync(basePath)) {
                changes.push({ path: childRel, type: 'added' });
            }
            _scanCloneSnapshotChanges(upperRoot, baseRoot, childRel, changes);
            continue;
        }

        if (!fs.existsSync(basePath)) {
            changes.push({ path: childRel, type: 'added' });
            continue;
        }

        if (_pathsDiffer(upperPath, basePath)) {
            changes.push({ path: childRel, type: 'modified' });
        }
    }
}

/**
 * An opaque directory marker (`.wh..wh..opq`) hides every lower entry of the
 * directory it lives in.  Report the lower entries the upper layer does not
 * contain — neither directly nor as a whiteout — as deleted.
 *
 * @param {string} upperRoot
 * @param {string} baseRoot
 * @param {string} rel   directory holding the marker, relative to the roots
 * @param {Array<object>} changes
 */
function _collectOpaqueDeletions(upperRoot, baseRoot, rel, changes) {
    const upperDir = path.join(upperRoot, rel);
    let lowerNames;
    try {
        lowerNames = fs.readdirSync(path.join(baseRoot, rel));
    } catch {
        return;   // no lower counterpart — nothing is hidden
    }

    for (const lowerName of lowerNames) {
        if (fs.existsSync(path.join(upperDir, lowerName))) continue;
        if (fs.existsSync(path.join(upperDir, WH_PREFIX + lowerName))) continue;
        changes.push({
            path: rel === '.' ? lowerName : path.join(rel, lowerName),
            type: 'deleted',
        });
    }
}

/**
 * Detect kernel overlay whiteout: character device with rdev 0.
 * @param {string} filePath
 * @returns {boolean}
 */
function _isKernelWhiteout(filePath) {
    try {
        const st = fs.lstatSync(filePath);
        return st.isCharacterDevice?.() && st.rdev === 0;
    } catch {
        return false;
    }
}

function _readCloneManifest(upperRoot) {
    const manifestPath = path.join(
        path.dirname(upperRoot),
        MANIFEST_DIR,
        `${path.basename(upperRoot)}${MANIFEST_SUFFIX}`,
    );

    if (!fs.existsSync(manifestPath) || !fs.statSync(manifestPath).isFile()) {
        return null;
    }

    return fs.readFileSync(manifestPath, 'utf8')
        .split(/\r?\n/u)
        .filter(Boolean);
}

function _hasDeletedAncestor(relPath, deletedAncestors) {
    let current = path.dirname(relPath);
    while (current && current !== '.') {
        if (deletedAncestors.has(current)) return true;
        current = path.dirname(current);
    }
    return false;
}

function _pathsDiffer(upperPath, basePath) {
    const upperStat = fs.lstatSync(upperPath);
    const baseStat = fs.lstatSync(basePath);

    if (upperStat.isSymbolicLink() || baseStat.isSymbolicLink()) {
        return upperStat.isSymbolicLink() !== baseStat.isSymbolicLink() ||
            fs.readlinkSync(upperPath) !== fs.readlinkSync(basePath);
    }

    if (upperStat.isFile() && baseStat.isFile()) {
        if (upperStat.size !== baseStat.size) return true;
        return !fs.readFileSync(upperPath).equals(fs.readFileSync(basePath));
    }

    return upperStat.isDirectory() !== baseStat.isDirectory();
}

/**
 * Format a list of changes as a human-readable summary.
 *
 * @param {Array<{ path: string, type: 'added'|'modified'|'deleted' }>} changes
 * @returns {string}
 */
export function formatChanges(changes) {
    if (changes.length === 0) return 'No changes detected.\n';
    return changes
        .map(c => `${c.type === 'added' ? 'A' : c.type === 'modified' ? 'M' : 'D'}\t${c.path}`)
        .join('\n') + '\n';
}

File: sdk/py/src/boxsh_py/changes.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from stat import S_ISCHR
from typing import List, Literal
import filecmp
import os

WhiteoutType = Literal["added", "modified", "deleted"]
_WHITEOUT_PREFIX = ".wh."
_MANIFEST_SUFFIX = ".manifest"


@dataclass(frozen=True)
class Change:
    path: str
    type: WhiteoutType


def get_changes(*, upper: str, base: str) -> List[Change]:
    manifest = _read_clone_manifest(Path(upper))
    if manifest is not None:
        return _get_clone_snapshot_changes(Path(upper), Path(base), manifest)

    changes: List[Change] = []
    _scan_changes(Path(upper), Path(base), Path("."), changes)
    return sorted(changes, key=lambda change: change.path)


def _get_clone_snapshot_changes(upper_root: Path, base_root: Path, manifest_paths: List[str]) -> List[Change]:
    changes: List[Change] = []
    deleted_ancestors: set[str] = set()

    for rel_path in manifest_paths:
        if _has_deleted_ancestor(rel_path, deleted_ancestors):
            continue
        if (upper_root / rel_path).exists():
            continue
        deleted_ancestors.add(rel_path)
        changes.append(Change(path=rel_path, type="deleted"))

    _scan_clone_snapshot_changes(upper_root, base_root, Path("."), changes)
    return sorted(changes, key=lambda change: change.path)


def _scan_changes(upper_root: Path, base_root: Path, rel: Path, changes: List[Change]) -> None:
    upper_dir = upper_root / rel
    if not upper_dir.exists() or not upper_dir.is_dir():
        return

    for entry in upper_dir.iterdir():
        child_rel = Path(entry.name) if rel == Path(".") else rel / entry.name
        upper_path = upper_root / child_rel

        if entry.name.startswith(_WHITEOUT_PREFIX):
            target_name = entry.name[len(_WHITEOUT_PREFIX):]
            target_rel = Path(target_name) if rel == Path(".") else rel / target_name
            changes.append(Change(path=target_rel.as_posix(), type="deleted"))
            continue

        if _is_kernel_whiteout(upper_path):
            changes.append(Change(path=child_rel.as_posix(), type="deleted"))
            continue

        if upper_path.is_dir():
            base_path = base_root / child_rel
            if not base_path.exists():
                changes.append(Change(path=child_rel.as_posix(), type="added"))
            _scan_changes(upper_root, base_root, child_rel, changes)
            continue

        if upper_path.is_file():
            base_path = base_root / child_rel
            changes.append(
                Change(path=child_rel.as_posix(), type="modified" if base_path.exists() else "added")
            )


def _scan_clone_snapshot_changes(upper_root: Path, base_root: Path, rel: Path, changes: List[Change]) -> None:
    upper_dir = upper_root / rel
    if not upper_dir.exists() or not upper_dir.is_dir():
        return

    for entry in upper_dir.iterdir():
        child_rel = Path(entry.name) if rel == Path(".") else rel / entry.name
        upper_path = upper_root / child_rel

        if entry.name.startswith(_WHITEOUT_PREFIX):
            target_name = entry.name[len(_WHITEOUT_PREFIX):]
            target_rel = Path(target_name) if rel == Path(".") else rel / target_name
            changes.append(Change(path=target_rel.as_posix(), type="deleted"))
            continue

        if _is_kernel_whiteout(upper_path):
            changes.append(Change(path=child_rel.as_posix(), type="deleted"))
            continue

        base_path = base_root / child_rel

        if upper_path.is_dir():
            if not base_path.exists():
                changes.append(Change(path=child_rel.as_posix(), type="added"))
            _scan_clone_snapshot_changes(upper_root, base_root, child_rel, changes)
            continue

        if not base_path.exists():
            changes.append(Change(path=child_rel.as_posix(), type="added"))
            continue

        if _paths_differ(upper_path, base_path):
            changes.append(Change(path=child_rel.as_posix(), type="modified"))


def _is_kernel_whiteout(file_path: Path) -> bool:
    try:
        stat_result = os.lstat(file_path)
    except OSError:
        return False
    return S_ISCHR(stat_result.st_mode) and os.major(stat_result.st_rdev) == 0 and os.minor(stat_result.st_rdev) == 0


def _read_clone_manifest(upper_root: Path) -> List[str] | None:
    manifest_path = upper_root.parent / ".boxsh" / f"{upper_root.name}{_MANIFEST_SUFFIX}"
    if not manifest_path.exists() or not manifest_path.is_file():
        return None
    return [line for line in manifest_path.read_text(encoding="utf-8").splitlines() if line]


def _has_deleted_ancestor(rel_path: str, deleted_ancestors: set[str]) -> bool:
    current = Path(rel_path).parent.as_posix()
    while current and current != ".":
        if current in deleted_ancestors:
            return True
        current = Path(current).parent.as_posix()
    return False


def _paths_differ(upper_path: Path, base_path: Path) -> bool:
    if upper_path.is_symlink() or base_path.is_symlink():
        return upper_path.is_symlink() != base_path.is_symlink() or os.readlink(upper_path) != os.readlink(base_path)

    if upper_path.is_file() and base_path.is_file():
        return not filecmp.cmp(upper_path, base_path, shallow=False)

    return upper_path.is_dir() != base_path.is_dir()


def format_changes(changes: List[Change]) -> str:
    if not changes:
        return "No changes detected.\n"
    marker = {"added": "A", "modified": "M", "deleted": "D"}
    return "".join(f"{marker[change.type]}\t{change.path}\n" for change in changes)